gno.land/pkg/integration/testing_integration.go

```diff
@@ -0,0 +1,257 @@
+package integration
+
+import (
+	"context"
+	"fmt"
+	"hash/crc32"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gnolang/gno/tm2/pkg/bft/node"
+	"github.com/gnolang/gno/tm2/pkg/commands"
+	"github.com/gnolang/gno/tm2/pkg/crypto/keys"
+	"github.com/gnolang/gno/tm2/pkg/crypto/keys/client"
+	"github.com/gnolang/gno/tm2/pkg/log"
+	"github.com/jaekwon/testify/require"
+	"github.com/rogpeppe/go-internal/testscript"
+)
+
+// XXX: This should be centralize somewhere.
+const (
+	test1Addr = "g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5"
+	test1Seed = "source bonus chronic canvas draft south burst lottery vacant surface solve popular case indicate oppose farm nothing bullet exhibit title speed wink action roast"
+)
+
+type testNode struct {
+	*node.Node
+	logger      log.Logger
+	nGnoKeyExec uint // Counter for execution of gnokey.
+}
+
+func SetupGnolandTestScript(t *testing.T, txtarDir string) testscript.Params {
+	t.Helper()
+
+	cmd := exec.Command("go", "list", "-m", "-mod=mod", "-f", "{{.Dir}}", "github.com/gnolang/gno")
+	out, err := cmd.CombinedOutput()
+	require.NoError(t, err)
+
+	// `gnoRootDir` should point to the local location of the gno repository.
+	// It serves as the gno equivalent of GOROOT.
+	gnoRootDir := strings.TrimSpace(string(out))
+
+	// `gnoHomeDir` should be the local directory where gnokey stores keys.
+	gnoHomeDir := filepath.Join(t.TempDir(), "gno")
+
+	// `gnoDataDir` should refer to the local location where the gnoland node
+	// stores its configuration and data.
+	gnoDataDir := filepath.Join(t.TempDir(), "data")
+
+	var muNodes sync.Mutex
```

Comment:
Is a mutex really necessary?

As far as I can see, we're only changing it within this function, and not within a gno.land execution. So concurrent access should not happen...

Reply:
yes it is, It can be surprising but testscripts directory files are executed [concurrently](https://github.com/rogpeppe/go-internal/blob/ec119421bb970ac974ea4ac16d8b58ffd0d727fc/testscript/testscript.go#L309), removing mutex make the test panic with a datarace error, I added a mutex from there

Reply:
> Is a mutex really necessary?

Yes, it is. It can be a bit surprising, but testscript files are executed [concurrently](https://github.com/rogpeppe/go-internal/blob/ec119421bb970ac974ea4ac16d8b58ffd0d727fc/testscript/testscript.go#L309). Without this, I experienced a panic. I added a mutex from there.

Reply:
a-ha! thanks for the explanation, might be good to add it to the code above the mutex.

if we do ever need to run tests sequentially, we can call testscript.RunT with a custom type, having a no-op `Parallel()` function. I think it's OK if we are parallel by default though (especially as these are likely tests that involve some seconds of waiting, and if we have 100+ of them they're gonna take too long...)